hazelcast/src/main/java/com/hazelcast/instance/HazelcastInstanceFactory.java

```diff
@@ -29,4 +29,6 @@ HazelcastInstance newHazelcastInstance(Config config, String instanceName,
                                            NodeContext nodeContext) throws Exception;
 
     HazelcastInstance newHazelcastInstance(Config config) throws Exception;
+
+    HazelcastInstanceProxy newHazelcastInstanceProxy(HazelcastInstanceImpl hazelcastInstance);
```

Comment:
yes, I know. My point is that adding impl class to interface seems not good way to me. 

Not sure but maybe you can use `HazelcastInstance` in the interface and make a type check and cast in the instance factory.